### ISSUE TITLE
Implement a way to customize "trap behavior"

### ISSUE THREAD

**Issue by kyouko-taiga** [hylo-lang/hylo#1046]
We need a way to explain how a program can trap without making a system call when it is compiled with `--no-std`. We should probably take inspiration from Rust's [#[panic_handler]](https://doc.rust-lang.org/nomicon/panic-handler.html#panic_handler).

**Comment by kyouko-taiga** [comment]
I propose the following design:
```
// In Library/Hylo/Core/Assert.hylo
@forward("Hylo.fatal_error")
public fun fatal_error(_ message: String = "", file: String = #file, line: Int = #line) -> Never

// In Library/Hylo/Assert+LibC.hylo
@definition("Hylo.fatal_error")
public fun fatal_error(_ message: String = "", file: String = #file, line: Int = #line) -> Never {
  print(file + ":" + line + ": fatal error: " + message)
  Builtin.trap()
}
```
The `@forward` attribute lets one declare a function without definition, just like in C++. The argument of the attribute is the linking name of the function. The `@definition` attribute lets you declare a function (and its definition) that will be linked to a forward declaration. All `@forward` functions must have exactly one corresponding `@definition` at link time, matching its linking name.

In normal mode (i.e., `--no-std`), the forwarded `fatal_error` function is linked to the definition in `Assert+LibC.hylo`, which calls `Builtin.trap()`. This built-in function corresponds to LLVM's `trap`, which may call `abort` from libc on some targets ([docs](https://llvm.org/docs/LangRef.html#llvm-trap-intrinsic)).

When one uses `--no-std`, only the contents of `Library/Hylo/Core` is loaded as part of the standard library, meaning that the default definition in `Assert+LibC.hylo` is not present. One must define `fatal_error` in their own application, e.g.:
```
// In MyDriver.hylo
@definition("Hylo.fatal_error")
public fun fatal_error(_ message: String = "", file: String = #file, line: Int = #line) -> Never {
  print(file + ":" + line + ": fatal error: " + message)
  while true() {}
}
```

**Comment by dabrahams** [comment]
I understand what you're getting at, but would prefer not to use the term “forward,” which implies some kind of order-dependence.  Maybe just "declare" or "declaration_only?" I think this ought to be considered in the same breath as FFI declarations; there's probably a unified mechanism here.  I also wonder why we're writing the `Hylo.` in the declaration-without-definition, since that's the scope of all other declarations in Hylo/Core.

**Comment by kyouko-taiga** [comment]
> Maybe just "declare" or "declaration_only?

Good point about `forward`. I like `@declare` because it's more concise. But since we're talking about a niche feature it might be best to go for the more descriptive `@declaration_only`. Maybe we can also consider `@nodef`. No strong opinion on my side. 

>  I also wonder why we're writing the `Hylo.` in the declaration-without-definitions.

It's a form of C-style namespacing, with the advantage that we can use any character we like because we'll probably mangle them anyway.

But perhaps what you're getting at is that we could say that `@declare_only` doesn't need any identifier. Then we'd simply expect the user to write the corresponding `@define` with the qualified name of the declaration without definition. For example:

```
// In Library/Hylo/Core/Assert.hylo
@declare_only
public fun fatal_error(...) -> Never

// In MyDriver.hylo
@definition("Hylo.fatal_error")
public fun fatal_error(...) -> Never { ... }
```

I think that's a better approach.

> there's probably a unified mechanism here.

I had the same thought but FFIs require the compiler to emit some additional glue code. This feature on the other hand would just be a way to separate a declaration from its definition.

When we write `@ffi("foo") fun foo(x: Int) `, we're _declaring_ two functions:
- `void foo(Builtin.word)`, and
- `void some_mangled_name(Int*, Void*)`.

The latter contains the glue code that translates Hylo types to C types (and vice versa) using `ForeignConvertible`.

**Comment by dabrahams** [comment]
> ```
> // In Library/Hylo/Core/Assert.hylo
> @declare_only
> public fun fatal_error(...) -> Never
> 
> // In MyDriver.hylo
> @definition("Hylo.fatal_error")
> public fun fatal_error(...) -> Never { ... }
> ```

```
// In Library/Hylo/Core/Assert.hylo
@declare_only
public fun fatal_error(...) -> Never

// In MyDriver.hylo
extension Hylo {
  @definition // do we even need this attribute?
  public fun fatal_error(...) -> Never { ... }
}
```

Makes me think that if we don't want to support extending modules, maybe these customization point should all just be required to be type members.  You can always supply the free function that calls them


  ```
// In Library/Hylo/Core/Assert.hylo
public fun fatal_error(...) -> Never { Hooks.fatal_error() }

type Hooks {
  @declare_only
  static fun fatal_error(...)
}

// In MyDriver.hylo
extension Hylo.Hooks {
  @definition
  public fun fatal_error(...) -> Never { ... }
}
```

> FFIs require the compiler to emit some additional glue code

Just goes to show you that I still don't understand your FFI scheme.

**Comment by kyouko-taiga** [comment]
>  do we even need this attribute?

Maybe not if we go with your `extension Module` idea (which I like). One would only be allowed to provide definitions for `@declare_only` functions, meaning that the compiler would understand what you're doing.

> Makes me think that if we don't want to support extending modules, maybe these customization point should all just be required to be type members.

I like module extensions better if I'm only thinking about the use of the original post. It also has the advantage of making it syntactically clear that we're defining a symbol that should be mangled in the namespace of another module, which is how we can get rid of `@definition`.

 A generalization of the feature seems to overlap with the trait system. AFAICT, the only reason your `Hooks` is not a trait is that `Hylo.fatal_error` must have a way to resolve `Hooks` to some concrete type that isn't available when we compile the standard library.

But ... I don't hate the idea of being able to separate definitions from their declarations. I believe there is some value in the header/source model of C++.
